File: app/models/track_sheet.rb
class TrackSheet < ApplicationRecord
  belongs_to :chapter, touch: true

  serialize :definition, coder: JSON

  def apply(updates)
    Marshal.load(Marshal.dump(definition || {})).tap do |copy|
      updates.each { |update| apply_update_to(copy, update) }
    end
  end

  private

  def apply_update_to(sheet, update)
    node = update['parent'] && find_parent(sheet, update['parent'])
    return unless node

    case update['action']
    when 'add' then
      node.update(update['child'])
    when 'update' then
      update['child'].each do |key, value|
        node[key]['value'] = value if node[key]
      end
    when 'rename' then
      update['child'].each do |old_name, new_name|
        node[new_name] = node.delete(old_name)
      end
    when 'remove' then
      update['child'].each { |child| node.delete(child) }
    else
      raise NotImplementedError, update['action']
    end
  end

  # Same as Hash#dig, but it also creates missing intermediate Hashes.
  def find_parent(sheet, path)
    path.inject(sheet) do |node, key|
      node[key] ||= {}
    end
  end
end

File: app/javascript/track-sheet.js
export default class {
  constructor(source) {
    this.source = structuredClone(source);
  }

  applyUpdates(updates) {
    for (let update of updates) {
      this.applyUpdate(update);
    }
  }

  applyUpdate(update) {
    if (update.action == 'add')
      return this.applyAddUpdate(update);
    if (update.action == 'update')
      return this.applyUpdateUpdate(update);
    if (update.action == 'rename')
      return this.applyRenameUpdate(update);
    if (update.action == 'remove')
      return this.applyRemoveUpdate(update);

    throw `not an acceptable update action: '${update.action}'`;
  }

  applyAddUpdate(update) {
    let node = this.findParent(update.parent);
    Object.assign(node, update.child);
  }

  applyUpdateUpdate(update) {
    let node = this.findParent(update.parent);
    if (!node) return;

    for (let prop in update.child) {
      if (node[prop]) {
        node[prop].value = update.child[prop];
      }
    }
  }

  applyRenameUpdate(update) {
    let node = this.findParent(update.parent);
    if (!node) return;

    for (let prop in update.child) {
      if (node[prop]) {
        const saved = node[prop];
        delete node[prop];
        node[update.child[prop]] = saved;
      }
    }
  }

  applyRemoveUpdate(update) {
    let node = this.findParent(update.parent);
    for (let child of update.child) {
      delete node[child];
    }
  }

  findParent(path) {
    let node = this.source;

    for (let name of path) {
      node[name] ||= {};
      node = node[name];
    }

    return node;
  }
}
